### PR TITLE
feat(qlog): log version_information on client

### PR DESCRIPTION
**Pull request by mxinden**
:wave: Hi there

This is Max, first time contributor.

Let me know if this tiny pull request is of some help. Would appreciate a review and pointers to what to tackle next.

Thank you for QUIC support in Firefox!

---

This commit adds support for the qlog [`version_information` QUIC event](https://quicwg.org/qlog/draft-ietf-quic-qlog-quic-events.html#name-version_information) on the client.

Depends on https://github.com/mozilla/neqo/pull/1504

Depends on https://github.com/cloudflare/quiche/pull/1684

Meta issue: https://github.com/mozilla/neqo/issues/528

This commit does not add server-side support. I was unsure where to add the `qlog::xxx` call. With my limited knowledge of the code base, I would place it in the [`TransportParametersHandler`](https://github.com/mozilla/neqo/blob/83735a88217a6b3a6a9d3cd5d9243040c5e41319/neqo-transport/src/tparams.rs#L633-L643), though as of today, it does not have access to the connection's `NeqoQlog`. Thoughts?

### CODE REVIEW

File: Cargo.toml
Comment (mxinden) [inline Cargo.toml:15]: To be removed once https://github.com/cloudflare/quiche/pull/1684 is merged and released.

Comment (martinthomson) [reply Cargo.toml:15]: We'll want to wait for that before we merge here.  For future reference, a draft PR is probably better for cases where you need to wait for upstream changes.  That might delay review, but we're generally happy to provide feedback on direction from a draft.

File: neqo-transport/src/qlog.rs
Comment (martinthomson) [inline neqo-transport/src/qlog.rs:155]: You can inline `ev_data` (it's less code overall).

**Review by martinthomson — APPROVED** [review]
For the server, look in `server.rs` for `fn accept_connection(...)`.  That is where a new connection is handled.  There shouldn't be a negotiated equivalent on the server, but the recording of failed version negotiation can be recorded.

For VN, you can make a call in `fn process_input(...)` right where it calls `PacketBuilder::version_negotiation(...)`, which is where the Version Negotiation packet is sent.  For that, you'll need to create a new QLog instance with `self.create_qlog_trace(attempt_key)`.  You will have to follow other code for that.  I probably wouldn't bother trying to reduce the number of logging events from the same remote address and DCID, because that just creates a cleanup hazard.

That won't catch all cases, but enough.

File: Cargo.toml
Comment (larseggert) [reply Cargo.toml:15]: https://github.com/cloudflare/quiche/pull/1684 is merged.

Comment (larseggert) [reply Cargo.toml:15]: @mxinden could you update the PR to not require the patched crate?

File: neqo-transport/src/qlog.rs
Comment (mxinden) [reply neqo-transport/src/qlog.rs]: Sounds good. Done in https://github.com/mozilla/neqo/pull/1505/commits/a5f7681671085ab0fa3f65bf8eb94927e6c44f4f.

File: Cargo.toml
Comment (mxinden) [reply Cargo.toml]: I am sorry for the delay here @larseggert.

With https://github.com/cloudflare/quiche/pull/1684 merged, I updated from `github.com/mxinden` to `github.com/cloudflare`. See https://github.com/mozilla/neqo/pull/1505/commits/5884b456e10c5e1f6a52e89af8c0fcacc86f07dd. I am assuming that we still want to block merging here until the patch in https://github.com/cloudflare/quiche/pull/1684 has been released to crates.io. See also https://crates.io/crates/qlog/versions. Let me know in case I am missing something.

Comment (larseggert) [reply Cargo.toml]: Yes, thanks for catching that. (I had assumed quiche would be uploaded to crates.io after the merge, but I guess not yet.)

Comment (mxinden) [reply Cargo.toml]: Patch section removed with https://github.com/mozilla/neqo/pull/1505/commits/4754645b1c6192fdf6a5d17a927acae2da18c17a with upgrade to `qlog` `v0.11.0` in https://github.com/mozilla/neqo/pull/1547. Resolving here.

**Comment by mxinden** [comment]
With https://github.com/mozilla/neqo/pull/1547 merged, all outstanding comments are addressed. @martinthomson or @larseggert could either of you take another look?

**Comment by martinthomson** [comment]
I opened #1549 for the server side.  I was waiting to see what you planned there, that's all.

**Comment by mxinden** [comment]
> I opened #1549 for the server side. I was waiting to see what you planned there, that's all.

Sorry, badly communicated from my end. I prepared the server side in https://github.com/mozilla/neqo/pull/1531.

**Pull request merged**
